// grunt-htmlbuild/Gruntfile.js
/*
 * grunt-htmlbuild
 * https://github.com/amekkawi/grunt-htmlbuild
 *
 * Copyright (c) 2013 Andre Mekkawi
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>'
      ],
      options: {
        jshintrc: '.jshintrc'
      }
    },

    // Before generating any new files, remove any previously-created files.
    clean: {

    },

    // Configuration to be run (and then tested).
    htmlbuild: {
      default_options: {
        options: {

        },
        files: {

        }
      },
      custom_options: {
        options: {

        },
        files: {

        }
      }
    },

    // Unit tests.
    nodeunit: {

    }

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'htmlbuild', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);

};
